SCClassLibrary/Common/GUI/PlusGUI/Control/NodeTreeView.sc

```diff
@@ -0,0 +1,256 @@
+NodeTreeView : View {
+
+	var server;
+	var <>displayControls = true;
+	var updateFunc, updater, <>refreshRate = 0.3;
+	var scroll, canvas;
+	var nodeTree, drawCommands, viewHeight, penPos;
+	var font, fontSize = 10, fontOverflow = 1.2, margins = 10;
+	var xIndex;
+	var string, truncValue = 0.001;
+
+	*new { |server|
+		^super.new.init(server)
+	}
+	
+	init { |parentServer|
+		server = parentServer;
+		font = Font.sansSerif(fontSize);
+		fontOverflow = (fontSize * fontOverflow).asInteger;
+		scroll = ScrollView();
+		canvas = UserView();
+		this.layout_(
+			VLayout()
+			.margins_(0)
+			.spacing_(0)
+			.add(scroll);
+		);
+		scroll.canvas_(canvas);
+		canvas.drawFunc_({ |view|
+			Pen.strokeColor_(Color.black);
+			nodeTree = server.getNodeTree;
+			if(drawCommands.isNil) {
+				Pen.fillColor_(Color.grey(0.75));
+				Pen.addRect(
+					Rect(
+						margins,
+						margins,
+						view.bounds.width - (margins * 2),
+						view.bounds.height - (margins * 2)
+					)
+				);
+				Pen.fillStroke;
+				Pen.stringCenteredIn(
+					"Server not running",
+					Rect(
+						0,
+						0,
+						this.bounds.width,
+						this.bounds.height
+					),
+					font,
+					Color.black
+				);
+			} {
+				penPos = margins;
+				drawCommands.do({ |command|
+					this.prDrawCommand(command, view);
+				});
+			};
+		});
+
+		updateFunc = {
+			updater = fork {
+				loop {
+					nodeTree = server.getNodeTree;
+					if(nodeTree.isNil) {
+						viewHeight = nil;
+						drawCommands = nil;
+					} {
+						this.prConstructDrawCommands;
+					};
+					{
+						if(viewHeight.isNil) {
+							viewHeight = this.bounds.height - 2;
+						};
+						canvas.fixedHeight_(viewHeight);
+						canvas.refresh;
+					}.defer;
+					refreshRate.wait;
+				}
+			}
+		};
+		updateFunc.value;
+
+		this.onClose_({
+			updater.stop;
+			CmdPeriod.remove(updateFunc);
+		});
+
+		this.onResize_({
+			if(drawCommands.isNil) {
+				canvas.fixedHeight_(this.bounds.height - 2);
+			};
+		});
+	}
+
+	prNodeAsDrawCommand { |node, parent, xIndex|
+		var command = ();
```

Comment:
In class library code, maybe an `IdentityDictionary` may be better here than an Event (unless Event functionality is needed).

Reply:
It isn't needed, that was typing laziness. I almost changed it to a `Dictionary`, would have been wrong anyway. Fixed.